File: daily/2019-11-08/46 Permutations.py

# order
# 

from itertools import permutations

class Solutions1:
    def permute(self, nums):
        return list(permutations(nums))

class Solutions2:
    def permute(self, nums):
        
        res = []
        self.dfs(nums, res, [])
        return res
    
    def dfs(self, nums, res, path):
        if not nums:
            res.append(path)
        
        else:
            for i in range(len(nums)):
                self.dfs(nums[:i] + nums[i+1:], res, path + [nums[i]])

class Solutions3:
    def permute(self, nums):

        visited = [0] * len(nums)
        res = []

        def dfs(path):
            if len(path) == len(nums):
                res.append(path)
            
            else:
                for i in range(len(nums)):
                    if not visited[i]:
                        visited[i] = 1
                        dfs(path+[nums[i]])
                        visited[i] = 0  # back-tracking
        dfs([])
        return res


from itertools import permutations
class Solution5:
    def permute(self, nums):
        
        # return list(permutations(nums))
    
    
        # back-tracking problem
        
        visited = [None] * len(nums)
        res = []
        path = []
        ans = []
     
        def dfs(nums, res, path, visited, ans):

            if len(path) == len(nums):
                res.append(path)
                ans.append(path)
                print(path)
                return
            else:
                for i in range(len(nums)):
                    if visited[i] == None:
                        visited[i] = True
                        path.append(nums[i])
                        dfs(nums, res, path, visited, ans)
                        path.pop()
                        visited[i] = None

        dfs(nums, res, path, visited, ans)

        return ans


class Solution4:
    def permute(self, nums):
        
        # return list(permutations(nums))
    
    
        # back-tracking problem
        
        visited = [None] * len(nums)
        res = []
        path = []
        self.dfs(nums, res, path, visited)
        return res
        
        
    def dfs(self, nums, res, path, visited):
        
        if len(path) == len(nums):
            res.append(path)
            return
        else:
            for i in range(len(nums)):
                if visited[i] == None:
                    visited[i] = True
                    path.append(nums[i])
                    self.dfs(nums, res, path, visited)
                    path.pop()
                    visited[i] = None

File: Leetcode First 200 JavaScript solution/46. Permutations.js
var permute = function(nums) {
    let res = [];
    dfs (nums, []);
    return res;

    function dfs(nums, path) {
        if (path.length == nums.length) {
            res.push(path.slice());
            return;
        }
        for (let i = 0; i < nums.length; i++) {
            if (path.includes(nums[i])) continue;
            path.push(nums[i]);
            dfs(nums, path);
            path.pop();
        }
    }
};

// use iteration to check the duplicate number
// this is pretty slow ! shit

var permutate = function(nums) {
    let res = [];
    let visited = Array(nums.length).fill(1);
    dfs(nums, visited, []);
    return res;

    function dfs(nums, visited, path){
        if (path.length === nums.length) {
            res.push(path.slice());
            return
        }

        for (let i = 0; i < nums.length; i++) {
            if (visited[i]){
                path.push(nums[i]);
                visited[i] = 0;
                dfs(nums, visited, path);
                visited[i] = 1;
                path.pop()
            }
        }
    }
}

// no need to be backtracking here

var permute = function(nums) {
    let ans = [];

    function dfs(nums, temp){
        if (temp.length === nums.length) ans.push(path);
        // iterate all the number
        for (let i = 0; i < nums.length; i++) {
            if (temp.includes(nums[i])) continue;
            dfs(nums, [...temp, nums[i]]);
        }
    }
    dfs(nums, []);
    return ans;
}
// N!